### PR TITLE
[v1alph2] Adding testing guidelines

### PR DESCRIPTION
**Pull request by maelk**
built on top of #187

### CODE REVIEW

File: docs/testing.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+defined in the file.
+
+A test file can also be created to contain integration tests, named after the
+original file `<filename>_integration_test.go`. This file should only contain
+integration tests, for example to test the controllers with the managers.
+
+## Mocking
+
+To mock the interface, we use mockgen and the gomock module to write the tests.
+The mock code should be placed in a folder `mocks` withing the folder containing
```
Comment (Xenwar) [inline docs/testing.md]: within

File: docs/testing.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+the mocked code. The mock files should be named after the original file
```
Comment (Xenwar) [inline docs/testing.md]: the mock file

File: docs/testing.md (hunk 1 above)
Comment (kashifest) [inline docs/testing.md]: within

File: docs/testing.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+using fake clients from controller-runtime package
+`sigs.k8s.io/controller-runtime/pkg/client/fake` or
+`"k8s.io/client-go/kubernetes/fake"`.
+
+## Testing Guidelines
+
+* The tests should use Ginkgo and Gomega, for example for the assertions, when
+  needed. Other libraries should not be used to keep uniformity in the tests.
+* The tests should be table-driven when there are multiple cases.
+* Table-driven tests should use subtests to improve the clarity of the results
```
Comment (Xenwar) [inline docs/testing.md]: improve clarify

**Comment by kashifest** [comment]
looks good to me

File: docs/testing.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+code and the developers should verify that the code coverage did not decrease
+due to their PR. The generated code should be excluded from the coverage (files
+starting with `zz_generated.`)
+
+In order to run the tests and get the cover profile, you can run
+
+```sh
+make test
+```
+
```
Comment (Jaakko-Os) [inline docs/testing.md]: This could be added:
For more detailed unit test coverage results, run
make unit<tab>
to see available options

**Comment by maelk** [comment]
/hold
This should not go in before #187 is merged

File: controllers/baremetalcluster_controller_test.go [hunk 5]
```diff
@@ -56,7 +56,7 @@ func TestReconcileNormal(t *testing.T) {
 			// Assert that Bar() is invoked.
```
Comment (Xenwar) [inline controllers/baremetalcluster_controller_test.go]: replace Bar with the relevant function/method being invoked.
same in other places

File: docs/testing.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+This document outlines the testing strategy applied in CAPBM. All developers
+should follow those guidelines to ensure uniformity and quality in the tests.
+
+## Code coverage
+
+The minimum code coverage required for the `api`, `baremetal` and `controllers`
+folders is **80%**. Any PR introducing new code should include tests for the new
+code and the developers should verify that the code coverage did not decrease
+due to their PR. The generated code should be excluded from the coverage (files
```
Comment (Xenwar) [inline docs/testing.md]: Generated codes should be excluded ...

File: docs/testing.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+make test
+```
+
+or
+
+```sh
+make unit
+```
+
+You can use the auto-completion of the make command to see the available
```
Comment (Xenwar) [inline docs/testing.md]: Take this to before 'make unit' or after 'hack/unit.sh'

File: docs/testing.md (hunk 6 above)
Comment (maelk) [reply docs/testing.md]: for me it should stay singular. we are referring to the code in general, it is not countable. or then maybe pieces of code ?

Comment (Xenwar) [reply docs/testing.md]: oh. Agree. 
Generated code ...
I was focusing on removing the specific 'The generated code ...'

**Pull request merged**